File: src/main/java/com/elpassion/whyrx/CallbackExample.java
package com.elpassion.whyrx;

public class CallbackExample {

    void calculate(Double input, final Callback<Double> onSuccess, final Callback<Throwable> onError) {
        calculateLog10(input, result -> calculateLog10(result, onSuccess, onError), onError);
    }

    private void calculateLog10(Double input, Callback<Double> onSuccess, Callback<Throwable> onError) {
        try {
            onSuccess.call(log10(input));
        } catch (Exception e) {
            onError.call(e);
        }
    }

    Double log10(Double input) {
        double log10 = Math.log10(input);
        if (Double.isNaN(log10)) {
            throw new IllegalArgumentException();
        }
        return log10;
    }
}

File: src/main/java/com/elpassion/whyrx/CallbackExample.kt
package com.elpassion.whyrx

import java.util.concurrent.Executor

class CallbackExample(val executor: Executor) {

    fun calculate(input: Double, resultCallback: Callback<Double>, errorCallback: Callback<Exception>) {
        calculateLog10(input, Callback {
            calculateLog10(it, resultCallback, errorCallback)
        }, errorCallback)
    }

    fun calculateLog10(input: Double, resultCallback: Callback<Double>, errorCallback: Callback<Exception>) {
        executor.execute {
            try {
                resultCallback.call(log10(input))
            } catch (e: Exception) {
                errorCallback.call(e)
            }
        }
    }
}


